### PR TITLE
T17,T20,T14: Initial NLP model and API implementation

### LINKED ISSUE #14: Generate our own flight time data
After research done in #13, it was decided that we'll generate our own flight time data since there were no free APIs that we could use

### LINKED ISSUE #17: NLP model prototyping
The prototype for our NLP model should be able to do the following:

- Detect the language a user is speaking and reply in that language
- Give an appropriate response to airport related questions and redirect users if the question doesn't relate to the airport

### LINKED ISSUE #20: Backend API implementation

### PR DESCRIPTION
**Pull request by shahnitya23**
Closes #17, closes #20, and closes #14. This is the implementation for the NLP model and API. The code still requires optimization/refining, however, in terms of logic and functionality, it achieves the minimal requirements.

### CODE REVIEW

File: backend/src/views.py
Comment (Netra-Patel) [inline backend/src/views.py:127]: Can we add a description for each function for readbility?

**Review by Netra-Patel — COMMENTED** [review]
For each function let's add some comments for any future users. Other than that the code looks good!

**Review by Dennull — CHANGES_REQUESTED** [review]
There's a lot to unpack here. For example I think the folder structure of the Django project is messy right now. I also want to test the API using something like Postman to see if we can get a proper response after an external request has been made. It'd probably be easier if we get on a call to pair program this, let me know if that sounds good!

File: backend/api/views.py
Comment (Dennull) [inline backend/api/views.py:272]: Sometimes if the audio wasn't captured properly an exception gets thrown here, can we please look at handling that more gracefully? Maybe use a try except block to catch that exception

image

**Review by Dennull — CHANGES_REQUESTED** [review]

File: backend/api/chatbotmodel.h5
Comment (shahnitya23) [reply backend/api/chatbotmodel.h5:1]: That is definitely possible, it would help clean up the folder. We would have to re-designate the file path in the code.

File: backend/api/urls.py
Comment (shahnitya23) [reply backend/api/urls.py:8]: This will be done internally. I had the training of the model in a different URL just for the purpose of testing out whether the functions I made in Django worked properly and communicated with the 'faux' frontend setup.

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py:20]: I played around with the earlier and that line of code can be replaced with `nltk.download('punkt')`.

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py:34]: I'm open to that. I agree with it making `views.py` a little bloated so definitely can give it a go.

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py:58]: Yes, I will need to change the places with `Intent_English.json` to `intent_english.json`.

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py:189]: Sure, we can set up a meeting to do this. I have played around a little with audio files, however, its implements from the frontend is something I don't know much about.

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py:275]: I would need to play around with this. In terms of variable storage and storage optimization, I can maybe try to delete the variable after its served its function instead.

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py:297]: I was looking into this. I am having a little of bit trouble customizing it for the various languages. I will look into the further.

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py:277]: Yes, I have been looking into it - similar to the comment I made below.

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py:355]: For sure. This was just for prototyping purposes.

**Comment by shahnitya23** [comment]
> For each function let's add some comments for any future users. Other than that the code looks good!

Function purposes and line-by-line comments have been made to `views.py`.

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py]: Updated in the push in my recent push.

File: backend/api/views.py
Comment (Dennull) [reply backend/api/views.py:45]: Following up on this, do you still want to do this?

File: backend/api/views.py
Comment (Dennull) [reply backend/api/views.py:244]: Let's leave this for now until we merge, we'll do some more research and revisit it after

File: backend/api/views.py
Comment (Dennull) [reply backend/api/views.py:399]: Okay, it's not that big of a deal, more of a nitpick for a cleaner syntax. I'll leave it up to you if you want to make the change or not

File: backend/api/views.py
Comment (Dennull) [reply backend/api/views.py:431]: We'll take care of more detailed error handling later, but let's change this to be the `Response` object for now so we don't forget

File: backend/api/views.py
Comment (Dennull) [inline backend/api/views.py:288]: Can you please explain to me what `test_val` is used for?

File: backend/api/views.py
Comment (Dennull) [inline backend/api/views.py:335]: Do we still need this? What if a language that isn't supported is selected, is that possible?

File: backend/api/views.py
Comment (Dennull) [inline backend/api/views.py:356]: What's the difference between "invalid audio" here, and "no audio" in `capture_and_recognize`? Do we need multiple try except blocks, or can we move this try except block into the function?

File: backend/api/views.py
Comment (Dennull) [inline backend/api/views.py:324]: Should we set this for other languages too? Or is this language the only one that sounded to fast/slow by default

**Review by Dennull — CHANGES_REQUESTED** [review]
Looking better! Just a few requests for explanations and some suggestions

File: backend/api/views.py
Comment (Dennull) [reply backend/api/views.py]: Nice!

File: backend/api/chatbotmodel.h5
Comment (shahnitya23) [reply backend/api/chatbotmodel.h5:1]: I'm open to it, we will need to be careful with reassigning the file paths.

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py:36]: I can give it a go this week and get back to you regarding how it works for me. I did start testing with this and I think it's doable.

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py:397]: I don't quite follow on `Response` object. Do you mean to replace `HttpResponse` with `Response`?

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py:337]: We don't need that part of the code. I was testing something and was going to come back to it - however, I will most likely get rid of it in the coming days. 

If the language isn't supported I can make changes to the main function to accommodate for that language. I have made it so that this would be an easy process if required (only a couple of code line changes).

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py]: The idea was that 'invalid audio' would be for audio it can't decipher and 'no audio' would be for when no audio was fed. I can play around with this and see if I can just make it into one instead.

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py:326]: Yes, so Spanish was the only language (from English and French) that sounded too fast, hence, I had to change the speed of speech.

File: backend/api/views.py
Comment (shahnitya23) [reply backend/api/views.py]: I cleaned up the `view.py` file quite a bit. The functionality remains the same and works well.

File: backend/api/chatbotmodel.h5
Comment (Dennull) [reply backend/api/chatbotmodel.h5:1]: We can do this later, let's get it merged and working first

File: backend/api/views.py
Comment (Dennull) [reply backend/api/views.py]: Yeah, it looks like you've done that in the most recent push so nice!

File: backend/api/management/commands/import_flights.py
Comment (Dennull) [inline backend/api/management/commands/import_flights.py:13]: This is hard-coded right now, before we merge can we fix it so that it works on everyone's machines?

File: backend/api/model_functions.py
Comment (Dennull) [inline backend/api/model_functions.py:212]: For my own clarity, can you please explain what `test_val` does to me?

**Review by Dennull — APPROVED** [review]

**Pull request merged**
